blob/azureblob/azureblob.go

```diff
@@ -162,38 +165,87 @@ type lazyCredsOpener struct {
 }
 
 func (o *lazyCredsOpener) OpenBucketURL(ctx context.Context, u *url.URL) (*blob.Bucket, error) {
-	o.init.Do(func() {
-		// Use default credential info from the environment.
-		// Ignore errors, as we'll get errors from OpenBucket later.
-		accountName, _ := DefaultAccountName()
-		accountKey, _ := DefaultAccountKey()
-		clientId, _ := DefaultClientId()
-		sasToken, _ := DefaultSASToken()
-		storageDomain, _ := DefaultStorageDomain()
-		isCDN, _ := DefaultIsCDN()
-		protocol, _ := DefaultProtocol()
-
-		isMSIEnvironment := adal.MSIAvailable(ctx, adal.CreateSender())
-		opts := Options{
-			StorageDomain: storageDomain,
-			Protocol:      protocol,
-			IsCDN:         isCDN,
-		}
+	// Use default credential info from the environment.
```

Comment:
I've taken some time to think about it and haven't been able to figure out a better way.

I've basically renamed lazyCredsOpener to routerCredsOpener. It is possible to encapsulate that logic into `func (o *URLOpener) OpenBucketURL` but that would make it super-complex.

I wouldn't keep `init.Do` either, because intuitively users of the library would expect to use multiple Azure accounts (subscriptions in Azure terms) if they supply different `subscription_id` parameters.

Additionally, just for extra justification of the added complexity and major changes:
Currently an azure blob URL looks like `azblob://container`  that doesn't provide full information about the object you want to open, because containers live in storage accounts, which live in subscriptions. So that information has to be passed using different ways.
After the PR it will look like `azblob://container?storage_account=account&subscription_id=xxx-xxx-xxx` which is enough to figure out how to access the storage container.

Reply:
Hi ekini,  thanks for the PR and thinking about this.

However, I think I'm going to stick with my intuition above and say that we don't want to support opening multiple storage accounts via changes in the URL. We've avoided putting things like that in URLs on purpose, and instead used environment variables for the most part.

Here's an Azure example from `secrets` for using the Azure CLI:
https://github.com/google/go-cloud/blob/master/secrets/azurekeyvault/akv.go#L97

It uses an environment variable to toggle it on and off.

If you want/need to connect to different storage accounts/subscriptions, you can either use the regular openers, or create custom `URLOpener`s (one for each account, by creating the appropriate `Pipeline`).